src/openrct2/network/Http.cpp

```diff
@@ -14,313 +14,214 @@
+    CURLcode curl_code = curl_easy_perform(curl);
+    if (curl_code != CURLE_OK)
     {
-        http_response_t * response = http_request(req);
-        callback2(response);
-    }, std::move(request2), callback);
-    thread.detach();
-}
+        throw std::runtime_error("Failed to perform request");
+    }
 
-void http_request_dispose(http_response_t *response)
-{
-    if (response->type == HTTP_DATA_JSON && response->root != nullptr)
-        json_decref(response->root);
-    else if (response->type == HTTP_DATA_RAW && response->body != nullptr)
-        free(response->body);
+    // gets freed by curl_easy_cleanup
+    char * content_type;
+    long   code;
+    curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &code);
+    curl_easy_getinfo(curl, CURLINFO_CONTENT_TYPE, &content_type);
+    res.status       = static_cast<Status>(code);
+    res.content_type = std::string(content_type);
 
-    free(response);
+    return res;
 }
 
-const char *http_get_extension_from_url(const char *url, const char *fallback)
+void DoAsync(const Request & req, std::function<void(Response & res)> fn)
 {
-    const char *extension = strrchr(url, '.');
-
-    // Assume a save file by default if no valid extension can be determined
-    if (extension == nullptr || strchr(extension, '/') != nullptr) {
-        return fallback;
-    } else {
-        return extension;
-    }
+    auto thread = std::thread([=]() {
+        Response res;
+        try
+        {
+            res = http::Do(req);
+        }
+        catch (std::exception & e)
+        {
+            res.error = e.what();
+            return;
+        }
+        fn(res);
+    });
+    thread.detach();
 }
 
-size_t http_download_park(const char * url, void * * outData)
+size_t download_park(const char * url, void ** outData)
 {
     // Download park to buffer in memory
-    HttpRequest2 request;
-    request.Url = url;
-    request.Method = "GET";
-    request.Type = HTTP_DATA_NONE;
-
-    http_response_t *response = http_request(request);
-
-    if (response == nullptr || response->status_code != 200) {
-        Console::Error::WriteLine("Failed to download '%s'", request.Url.c_str());
-        if (response != nullptr) {
-            http_request_dispose(response);
-        }
+    Request request;
+    request.url    = url;
+    request.method = Method::GET;
 
+    Response res;
+    try
+    {
+        res = Do(request);
+        if (res.status != Status::OK)
+            std::runtime_error("bad http status");
+    }
+    catch (std::exception & e)
+    {
+        Console::Error::WriteLine("Failed to download '%s', cause %s", request.url.c_str(), e.what());
         *outData = nullptr;
         return 0;
     }
 
-    size_t dataSize = response->size - 1;
-    void * data = malloc(dataSize);
-    if (data == nullptr) {
-        dataSize = 0;
+    size_t dataSize = res.body.size() - 1;
+    void * data     = malloc(dataSize);
```

Comment:
`Memory::Allocate`

Reply:
Don't use Memory::Allocate — we're moving away from it entirely, cf. #7066

Reply:
Hmm, there's https://github.com/OpenRCT2/OpenRCT2/blob/9a362a8aa382334f04b88938490990d5adb00021/src/openrct2/core/MemoryStream.cpp#L59 which is not touched by the PR you mentioned.

@IntelOrca ?

Reply:
We are all C++ now so the method should be changed to return an `std::vector`.

Reply:
Didn't want to touch this but I'm fine which changing the function signature to be more C++ like.

Reply:
I've looked closer at this and I'm not sure why a explicit `download_park` function is required. It's a normal download which as well passed the URL. I've removed it and put the few lines into context.

Reply:
Wasn't it related somehow to solving the problem of having to download a park first to detect whether it really is a park (sv6)? If you search git history for `tmpnam` maybe you will something useful around there.